test-data/unit/check-optional.test

```diff
@@ -832,3 +832,133 @@ main:4: error: Argument 1 to "asdf" has incompatible type "List[str]"; expected
 main:4: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
 main:4: note: Consider using "Sequence" instead, which is covariant
 [builtins fixtures/list.pyi]
+
+[case testOptionalBackwards]
+from typing import Any, Optional
+
+def f1(b: bool) -> Optional[int]:
+    if b:
+        z = 10
+        reveal_type(z)  # N: Revealed type is "builtins.int"
+    else:
+        z = None
+        reveal_type(z)  # N: Revealed type is "None"
+    reveal_type(z)  # N: Revealed type is "Union[builtins.int, None]"
+    return z
+
+def f2(b: bool) -> int:
+    if b:
+        z = 10
+    else:
+        z = None
+    return z  # E: Incompatible return value type (got "Optional[int]", expected "int")
+
+def f3(b: bool) -> int:
+    # XXX: This one is a little questionable! Maybe we *do* want to allow this?
```

Comment:
I think this would be nice to have, but it probably should be discussed more (esp. since there are so few cases where the type is different from the first assignment)

Reply:
Yeah, and if we want to allow this, this can be implemented in a separate PR.